Refactor input source switching logic to combine MRU toggle with full…

mixalbl4-127 opened the pull request:
Refactor input source switching logic to combine MRU toggle with full cycle

Previously Quick Lang Switch always cycled through all input sources in order,
replacing GNOME’s default MRU (most recently used) behavior. This caused an
unintuitive experience for users who expect Alt+Space to toggle only between
the last two layouts, while still allowing cycling through the full list when
Alt is held and Space is pressed repeatedly.

Changes:
- Reimplemented `_quickSwitchLayouts` to preserve MRU toggle on the first key
  press (switch to the previously used source).
- Added a short press “session” counter: subsequent presses while Alt is held
  advance through the full list of layouts (forward/backward).
- Kept `activate(true)` to bypass the popup and prevent PiP focus issues.
- Retained reverse cycling with Shift.

This restores GNOME’s natural MRU switching while still fixing the PiP focus
bug and allowing intentional full cycling when desired.

mixalbl4-127 committed:
Refactor input source switching logic to combine MRU toggle with full cycle

Previously Quick Lang Switch always cycled through all input sources in order,
replacing GNOME’s default MRU (most recently used) behavior. This caused an
unintuitive experience for users who expect Alt+Space to toggle only between
the last two layouts, while still allowing cycling through the full list when
Alt is held and Space is pressed repeatedly.

Changes:
- Reimplemented `_quickSwitchLayouts` to preserve MRU toggle on the first key
  press (switch to the previously used source).
- Added a short press “session” counter: subsequent presses while Alt is held
  advance through the full list of layouts (forward/backward).
- Kept `activate(true)` to bypass the popup and prevent PiP focus issues.
- Retained reverse cycling with Shift.

This restores GNOME’s natural MRU switching while still fixing the PiP focus
bug and allowing intentional full cycling when desired.

ankostis commented:
Thank you very much, that was something i always wanted to achieve.
I don't have the gnome expertise to judge your feature, a trust you that it runs ok. 

Would it be to much to ask you to reformat your code with x4 spaces indentation, to avoid excess diffs with existing code?

Alternatively, if you think x2 spaces is worth keeping (we don't have that many nested blocks, do we?)
you may rewrite this PR and split it in 3 commits: 

- one commit to reformat pre-existing code with x2 spaces and other styling like trimming spaces at EOL,
   
- another commit create the file `.git-blame-ignore-rev` file and add the previous commit into it with content like this:
  ```
  # .git-blame-ignore-revs
  # styling: converted identation from 4 -> 2 spaces.
  0123456789abcd...
  ``` 

  and a finally
- a  commit to just add your feature.

ankostis commented:
I see it has conflicts.
Is there a reason why you choose to base your changes on an older version of this extension v8 (Jul 2023)?

ankostis commented:
(answering to myself) i guess it's because of [my last comment in #7](https://github.com/ankostis/gnome-shell-quick-lang-switch/issues/7#issuecomment-1658940588).

The problem is that the source has evolved significantly from v8 and cannot merge it.

mixalbl4-127 committed:
pairwise language switching with reversible Cyrillic set for GNOME 42.9

Reworked the input source switching logic to support paired language cycling 
(Latin ↔ Cyrillic) while preserving the extension’s core behavior of hiding 
the GNOME input source OSD popup. 

New behavior:
- Normal switch (switch-input-source) toggles between Latin (English) and 
  the currently active Cyrillic layout (Russian or Ukrainian).
- Reverse switch (switch-input-source-backward) changes the Cyrillic layout 
  to the next available one (Ru ↔ Ua) and activates it as the new paired 
  layout with English.
- Works regardless of the current layout; reversing always ensures a Cyrillic 
  layout is active and the new En↔Cyr pair is established.

This makes switching faster and more predictable when using multiple Cyrillic 
layouts, reducing unnecessary cycling and keeping the UX clean (no popup). 
Fully compatible with Ubuntu 22.04 GNOME Shell 42.9 and preserves original 
popup suppression behavior.
